File: 117. Populating Next Right Pointers in Each Node II/main.py
class Node:
    def __init__(
        self,
        val: int = 0,
        left: "Node" = None,
        right: "Node" = None,
        next: "Node" = None,
    ):
        self.val = val
        self.left = left
        self.right = right
        self.next = next


"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""


class Solution:
    def connect(self, root: "Node") -> "Node":
        if not root:
            return None

        queue = [root]

        while len(queue) != 0:
            size = len(queue)
            prev = None

            for i in range(size):
                node = queue.pop(0)

                if prev:
                    prev.next = node

                prev = node

                if node.left:
                    queue.append(node.left)

                if node.right:
                    queue.append(node.right)

        return root


"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""


class Solution:
    def getNext(self, node: "Node") -> "Node":
        while node:
            if node.left:
                return node.left

            if node.right:
                return node.right

            node = node.next

        return None

    def connect(self, root: "Node") -> "Node":
        if not root:
            return None

        if root.left:
            if root.right:
                root.left.next = root.right
            else:
                root.left.next = self.getNext(root.next)

        if root.right:
            root.right.next = self.getNext(root.next)

        self.connect(root.right)
        self.connect(root.left)

        return root


"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""


class Solution:
    def connect(self, root: "Node") -> "Node":
        if not root:
            return None

        node = root

        while node:
            leftestNode = None
            nextNode = None

            while node:
                if node.left:
                    if not leftestNode:
                        nextNode = node.left
                        leftestNode = nextNode
                    else:
                        nextNode.next = node.left
                        nextNode = nextNode.next

                if node.right:
                    if not leftestNode:
                        nextNode = node.right
                        leftestNode = nextNode
                    else:
                        nextNode.next = node.right
                        nextNode = nextNode.next

                node = node.next

            node = leftestNode

        return root

File: 117. Populating Next Right Pointers in Each Node II/main.cpp
#include <queue>
#include <vector>

using namespace std;

class Node {
public:
    int val;
    Node *left;
    Node *right;
    Node *next;

    Node()
        : val(0)
        , left(nullptr)
        , right(nullptr)
        , next(nullptr) {}

    Node(int _val)
        : val(_val)
        , left(nullptr)
        , right(nullptr)
        , next(nullptr) {}

    Node(int _val, Node *_left, Node *_right, Node *_next)
        : val(_val)
        , left(_left)
        , right(_right)
        , next(_next) {}
};

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/
class Solution {
    Node *getNext(Node *node) {
        while (node) {
            if (node->left) {
                return node->left;
            }

            if (node->right) {
                return node->right;
            }

            node = node->next;
        }

        return nullptr;
    }

public:
    Node *connect(Node *root) {
        if (!root) {
            return nullptr;
        }

        if (root->left) {
            if (root->right) {
                root->left->next = root->right;
            } else {
                root->left->next = getNext(root->next);
            }
        }

        if (root->right) {
            root->right->next = getNext(root->next);
        }

        connect(root->right);
        connect(root->left);

        return root;
    }
};

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/
class Solution {
    static constexpr int ARRAY_SIZE = 6000;

public:
    Node *connect(Node *root) {
        if (!root) {
            return nullptr;
        }

        Node *array[ARRAY_SIZE] = {NULL};
        array[0] = root;
        int head = 0;
        int tail = 1;

        while (head < tail) {
            int next = tail;

            for (int i = head; i < next; i++) {
                if (array[i]->left) {
                    array[tail++] = array[i]->left;
                }

                if (array[i]->right) {
                    array[tail++] = array[i]->right;
                }

                if (i < next - 1) {
                    array[i]->next = array[i + 1];
                }
            }

            array[next - 1]->next = nullptr;
            head = next;
        }

        return root;
    }
};

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/
class Solution {
    static constexpr int ARRAY_SIZE = 6000;

public:
    Node *connect(Node *root) {
        if (!root) {
            return nullptr;
        }

        Node *node = root;

        while (node) {
            Node *leftestNode = nullptr;
            Node *nextNode = nullptr;

            while (node) {
                if (node->left) {
                    if (!leftestNode) {
                        nextNode = node->left;
                        leftestNode = nextNode;
                    } else {
                        nextNode->next = node->left;
                        nextNode = nextNode->next;
                    }
                }

                if (node->right) {
                    if (!leftestNode) {
                        nextNode = node->right;
                        leftestNode = nextNode;
                    } else {
                        nextNode->next = node->right;
                        nextNode = nextNode->next;
                    }
                }

                node = node->next;
            }

            node = leftestNode;
        }

        return root;
    }
};

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/
class Solution {
public:
    Node *connect(Node *root) {
        if (!root) {
            return nullptr;
        }

        queue<Node *> q;
        q.push(root);

        while (!q.empty()) {
            int size = q.size();
            Node *prev = nullptr;

            for (int i = 0; i < size; i++) {
                Node *node = q.front();
                q.pop();

                if (prev) {
                    prev->next = node;
                }

                prev = node;

                if (node->left) {
                    q.push(node->left);
                }

                if (node->right) {
                    q.push(node->right);
                }
            }
        }

        return root;
    }
};